File: Top_100_Liked_Questions/236._Lowest_Common_Ancestor_of_a_Binary_Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    // recursive post order traversal
    let ans = [null];
    let helper = function(node) {
        if (!node) {
            return false;
        }
        
        let left = helper(node.left);
        let right = helper(node.right);
        
        let mid = (node === p || node === q);
        
        if (mid + left + right >= 2) {
            ans[0] = node;
        }
        
        return mid || left || right;
    }
    
    helper(root);
    return ans[0];
}

var lowestCommonAncestor2 = function(root, p, q) {
    // find the index of the p and q as if the tree is in the array form
    if (root == null) {
        return null;
    }
    
    let p_idx = null;
    let q_idx = null;
    let queue = [[root, 0]];
    while (queue.length && (p_idx == null || q_idx == null)) {
        let [node, idx] = queue.shift();
        if (node.val === p.val) {
            p_idx = idx;
        } else if (node.val === q.val) {
            q_idx = idx;
        }
        
        if (node.left != null) {
            queue.push([node.left, idx*2+1]);
        }
        if (node.right != null) {
            queue.push([node.right, idx*2+2]);
        }
    }
    
    // get the parent index of the bigger one until the indices match
    while (p_idx != q_idx) {
        if (p_idx < q_idx) {
            q_idx = Math.floor((q_idx-1)/2);
        } else {
            p_idx = Math.floor((p_idx-1)/2);
        }
    }
    
    let lca_idx = p_idx;
    if (lca_idx === 0) {
        return root;
    }
    
    // with the lca index, get the path to the root node
    let path = [];
    while (lca_idx > 0) {
        path.push(lca_idx);
        lca_idx = Math.floor((lca_idx-1)/2);
    }
    
    // reverse traverse the path and return the node object
    let idx = 0;
    let ansNode = root;
    for (let p of path.reverse()) {
        if (p === idx*2+1) {
            ansNode = ansNode.left;
            idx = idx*2+1;
        } else {
            ansNode = ansNode.right;
            idx = idx*2+2;
        }
    }
    
    return ansNode;
};

File: Top_100_Liked_Questions/236._Lowest_Common_Ancestor_of_a_Binary_Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lca;

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        this.lca = null;
        this.helper(root, p, q);
        return this.lca;
    }

    public boolean helper(TreeNode node, TreeNode p, TreeNode q){
        if (node == null)
            return false;

        boolean left = helper(node.left, p, q);
        boolean right = helper(node.right, p, q);
        boolean either = (node == p) || (node == q);

        if ( (left ? 1 : 0) + (right ? 1 : 0) + (either ? 1 : 0) >= 2 )
            this.lca = node;

        return left || right || either;
    }
}


class Solution2 {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return this.helper(root, p, q).lca;
    }

    public Triplet helper(TreeNode node, TreeNode p, TreeNode q) {
        if (node == null)
            return new Triplet();

        Triplet tl = helper(node.left, p, q);
        Triplet tr = helper(node.right, p, q);

        if (tl.lca != null)
            return tl;
        if (tr.lca != null)
            return tr;

        boolean has_p = tl.has_p || tr.has_p || node == p;
        boolean has_q = tl.has_q || tr.has_q || node == q;
        TreeNode lca = has_p && has_q ? node : null;
        return new Triplet(has_p, has_q, lca);
    }
}

class Triplet {
    public boolean has_p;
    public boolean has_q;
    public TreeNode lca;

    public Triplet(boolean has_p, boolean has_q, TreeNode lca) {
        this.has_p = has_p;
        this.has_q = has_q;
        this.lca = lca;
    }

    public Triplet() {
        this.has_p = false;
        this.has_q = false;
        this.lca = null;
    }
}